File: sort_algorithm/count_sort.go
package sort_algorithm

import "fmt"

type CountSort []int

func (c CountSort) SortImpl() {
	if len(c) < 2 {
		return
	}
	min, max := c[0], c[0]
	for _, cc := range c {
		if cc < min {
			min = cc
		}
		if cc > max {
			max = cc
		}
	}
	d := max - min
	arr := make([]int, d+1)
	for _, cc := range c {
		arr[cc-min]++
	}
	j := 0
	for i, cc := range arr {
		for ; cc > 0; cc-- {
			c[j] = i + min
			j++
		}
	}
	fmt.Println(c)
}

File: java/sort_algorithm/impl/CountSort.java
package sort_algorithm.impl;

import sort_algorithm.Sort;

import java.util.Arrays;
import java.util.stream.IntStream;

public class CountSort implements Sort {
    @Override
    public void sort(int[] arr) {
        int max = IntStream.of(arr).max().getAsInt(), min = IntStream.of(arr).min().getAsInt();
        final int diff = max - min + 1;
        int[] counts = new int[diff];
        for (int a: arr) {
            counts[a-min]++;
        }
        arr = new int[arr.length];
        for (int i = 0, j = 0; i < counts.length; i++) {
            while(counts[i]-- > 0) {
                arr[j++] = i + min;
            }
        }
        System.out.println(Arrays.toString(arr));
    }
}
